src/main/java/com/marklogic/kafka/connect/source/SerializedQueryHandler.java

```diff
@@ -34,26 +35,50 @@ public SerializedQueryHandler(DatabaseClient databaseClient, Map<String, Object>
         } else {
             constraintColumnName = QueryHandlerUtil.sanitize(rawConstraintColumnName);
         }
+        rowLimit = (Integer) parsedConfig.get(MarkLogicSourceConfig.ROW_LIMIT);
     }
 
     @Override
     public PlanBuilder.Plan newPlan(String previousMaxConstraintColumnValue) {
-        currentSerializedQuery = injectConstraintIntoQuery(previousMaxConstraintColumnValue);
+        currentSerializedQuery = appendConstraintAndOrderByToQuery(userSerializedQuery, previousMaxConstraintColumnValue);
+        if (rowLimit > 0) {
+            currentSerializedQuery = appendLimitToQuery(currentSerializedQuery);
+        }
         logger.info("Serialized query: " + currentSerializedQuery);
         return databaseClient.newRowManager().newRawPlanDefinition(new StringHandle(currentSerializedQuery));
     }
 
-    protected String injectConstraintIntoQuery(String previousMaxConstraintColumnValue) {
-        String constrainedSerialized = userSerializedQuery;
-        if (previousMaxConstraintColumnValue != null) {
-            String constraintNodeString = "{\"ns\":\"op\", \"fn\":\"where\", \"args\":[{\"ns\":\"op\", \"fn\":\"gt\", \"args\":[{\"ns\":\"op\", \"fn\":\"col\", \"args\":[\"" + constraintColumnName + "\"]}, \"" + previousMaxConstraintColumnValue + "\"]}]}";
-            try {
-                ObjectNode constraintNode = (ObjectNode) mapper.readTree(constraintNodeString);
+    protected String appendLimitToQuery(String currentSerializedQuery) {
+        String limitNodeString = "{\"ns\": \"op\", \"fn\": \"limit\", \"args\": [" + rowLimit + "]}";
+        try {
+            ObjectNode limitNode = (ObjectNode) mapper.readTree(limitNodeString);
```

Comment:
Not worth changing now, but I think it would read a bit nicer to just build this JSON manually since it's simple enough, and it avoids any error from having to call readTree. Again, out of scope for this PR, just something to consider. The problem is only that Java is ugly with JSON strings - this would be much easier to read if we were writing JavaScript.